### ISSUE TITLE
Allow replaceRange to not add a history state

### ISSUE THREAD

**Issue by dcrousso** [codemirror/CodeMirror#3436]
When working with CSS autocompletions, it is a core feature that highlighting a particular completion will show the resulting style on the page for that element as if it were to be applied (for example, typing "position: abs", having "absolute" autocomplete, and then being able to see what the element with "position: absolute;" would look like).  In order to ensure that undo/redo is not disturbed because of this, however, it is necessary to ensure that the adding and removing of that replaced text does not affect the history of the text area.  As such, I think that another parameter should be added to replaceRange that, when true, will prevent that operation from adding a history event.

**Comment by marijnh** [comment]
CodeMirror's history is implemented in a way that only allows strict rollback. You can't undo a change without also undoing everything that happened after it. Thus, what you propose isn't possible with the current history implementation. So I'd suggest you find another solution. I didn't entirely understand your description of the use case, but couldn't you make the change outside of the editor, after getting out the text?

**Comment by dcrousso** [comment]
I think I may have explained my use case somewhat poorly.  Let me try again.
Basically, I have an autocomplete system implemented in my editor that shows a list of values that match the current word every time the user presses a key.  If the user has typed "wi", then a popup appears with the matches "width" and "window".  A key part of this system is that whenever the user hovers over one of the completions in the popup, the hovered item will replace what the user has currently typed (so if they hovered over "window" then the text in the editor will display "window" instead of "wi").  I am currently using replaceRange to accomplish this.  The issue arises if the user decides to move their mouse either out of the popup entirely (thus deselecting all of the completions) or to hover a different item (like "width" instead of "window").  When this happens, I am replacing the current text (which is "window" in this case since I already replaced the entered text) with the text of the new hovered item (which is "width").  This replacement causes an extra history state to be added and makes it such that if the user were to undo, they would see the text go from "width" to "window" to "wi", which is very weird.  What I am asking for is to give replaceRange another parameter to prevent it from adding to the history, similar to how some of the other commands have options to ignore history events (markText has addToHistory).  With this parameter, then I could make the completion replacements not add to the history (like when "wi" was replaced with "window" or when "window" was replaced with "width") until the user presses the enter key and commits the selected completion, ensuring that the undo history would show as "window" to "wi".  Please let me know if that makes more sense.  Thanks.

**Comment by marijnh** [comment]
> What I am asking for is to give replaceRange another parameter to prevent it from adding to the history,

I explained why that doesn't work.

Have you tried simply undoing your changes to cancel them? That will also pop them off the history. (Be sure to call `.changeGeneration(true)` before making a change to ensure it gets its own history event).

**Comment by dcrousso** [comment]
Here is a sample document that gives a preview of my issue.  If you start typing the word "window" into the CodeMirror area, hit enter before you type the letter "d", and then try to undo, the cursor will move around to the previous edit but will not remove the text that was added by replaceRange.
```html
<script src="codemirror.js"></script>
<link rel="stylesheet" href="codemirror.css">
<p>Type "window" in the editor</p>
<style>
.CodeMirror { border: 1px solid black; }
.completion-hint {
    color: red;
    opacity: 0.5;
}
</style>
<textarea id="code"></textarea>
<script>
    var from = {line: 0, ch: 0};
    var to = {line: 0};
    var marker = null;
    function hideCompletion() {
        if (marker) {
            var range = marker.find();
            if (range) {
                marker.clear();
                cm.replaceRange("", range.from, range.to, "+removeCompletion");
            }
            marker = null;
        }
    }
    var cm = CodeMirror.fromTextArea(document.getElementById("code"));
    cm.addKeyMap({
        "Enter": function() {
            hideCompletion();
            var cursor = cm.getCursor();
            cm.replaceRange("window", from, cursor);
        }
    });
    cm.on("change", function(cm, change) {
        if (change.origin === "+completion" || change.origin === "+removeCompletion")
            return;
        if (!change.origin || change.origin.charAt(0) !== "+") {
            hideCompletion();
            return;
        }
        hideCompletion();
        var cursor = cm.getCursor();
        cm.replaceRange("window", from, to, "+completion");
        cm.setCursor(cursor);
        marker = cm.markText(cursor, to, {className: "completion-hint"});
    });
</script>
```
I haven't used changeGeneration before, but from what I can see in the docs I am not sure why I would want to prevent the state after undo from being merged with other events.  Could you possibly explain a bit more what the purpose of changeGeneration is with undo?  Thanks.

**Comment by dcrousso** [comment]
If it helps at all, I tried adding a parameter to replaceRange to prevent it from adding to the history.  I am not sure that I did it correctly or according to your conventions, but this is the general idea.  Thanks again for the quick responses and comments.  Sorry if I am being a bother.
```diff
@@ -4253,19 +4253,21 @@
       to: change.to,
       text: change.text,
       origin: change.origin,
+      ignoreHistory: change.ignoreHistory,
       cancel: function() { this.canceled = true; }
     };
-    if (update) obj.update = function(from, to, text, origin) {
+    if (update) obj.update = function(from, to, text, origin, ignoreHistory) {
       if (from) this.from = clipPos(doc, from);
       if (to) this.to = clipPos(doc, to);
       if (text) this.text = text;
       if (origin !== undefined) this.origin = origin;
+      if (ignoreHistory) this.ignoreHistory = ignoreHistory;
     };
     signal(doc, "beforeChange", doc, obj);
     if (doc.cm) signal(doc.cm, "beforeChange", doc.cm, obj);
 
     if (obj.canceled) return null;
-    return {from: obj.from, to: obj.to, text: obj.text, origin: obj.origin};
+    return {from: obj.from, to: obj.to, text: obj.text, origin: obj.origin, ignoreHistory: obj.ignoreHistory};
   }
 
   // Apply a change to a document, and add it to the document's
@@ -4286,7 +4288,7 @@
     var split = sawReadOnlySpans && !ignoreReadOnly && removeReadOnlyRanges(doc, change.from, change.to);
     if (split) {
       for (var i = split.length - 1; i >= 0; --i)
-        makeChangeInner(doc, {from: split[i].from, to: split[i].to, text: i ? [""] : change.text});
+        makeChangeInner(doc, {from: split[i].from, to: split[i].to, text: i ? [""] : change.text, ignoreHistory: change.ignoreHistory});
     } else {
       makeChangeInner(doc, change);
     }
@@ -4295,7 +4297,8 @@
   function makeChangeInner(doc, change) {
     if (change.text.length == 1 && change.text[0] == "" && cmp(change.from, change.to) == 0) return;
     var selAfter = computeSelAfterChange(doc, change);
-    addChangeToHistory(doc, change, selAfter, doc.cm ? doc.cm.curOp.id : NaN);
+    if (!change.ignoreHistory)
+      addChangeToHistory(doc, change, selAfter, doc.cm ? doc.cm.curOp.id : NaN);
 
     makeChangeSingleDoc(doc, change, selAfter, stretchSpansOverChange(doc, change));
     var rebased = [];
@@ -4484,11 +4487,11 @@
     cm.display.selForContextMenu = null;
   }
 
-  function replaceRange(doc, code, from, to, origin) {
+  function replaceRange(doc, code, from, to, origin, ignoreHistory) {
     if (!to) to = from;
     if (cmp(to, from) < 0) { var tmp = to; to = from; from = tmp; }
     if (typeof code == "string") code = doc.splitLines(code);
-    makeChange(doc, {from: from, to: to, text: code, origin: origin});
+    makeChange(doc, {from: from, to: to, text: code, origin: origin, ignoreHistory: ignoreHistory});
   }
 
   // SCROLLING THINGS INTO VIEW
@@ -7321,10 +7324,10 @@
                         text: this.splitLines(code), origin: "setValue", full: true}, true);
       setSelection(this, simpleSelection(top));
     }),
-    replaceRange: function(code, from, to, origin) {
+    replaceRange: function(code, from, to, origin, ignoreHistory) {
       from = clipPos(this, from);
       to = to ? clipPos(this, to) : from;
-      replaceRange(this, code, from, to, origin);
+      replaceRange(this, code, from, to, origin, ignoreHistory);
     },
     getRange: function(from, to, lineSep) {
       var lines = getBetween(this, clipPos(this, from), clipPos(this, to));

```

**Comment by dcrousso** [comment]
Hiya.  Just wanted to know if you have thought about this any more.  Thanks!

**Comment by marijnh** [comment]
Well, I see that you implemented the thing that I assured you was not safe. That's fine, you might never hit a situation where it crashes your editor. I am still not interested in merging it in, though.
